File: app/src/main/java/se/jakob/knarkklocka/AlarmActivity.java
/*
 * Copyright (C) 2014 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package se.jakob.knarkklocka;

import android.app.AlarmManager;
import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.os.SystemClock;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.Chronometer;
import android.widget.TextView;

import java.util.Date;

import se.jakob.knarkklocka.data.Alarm;
import se.jakob.knarkklocka.data.AlarmActivityViewModel;
import se.jakob.knarkklocka.utils.TimerUtils;

import static android.app.AlarmManager.RTC_WAKEUP;
import static android.text.format.DateUtils.MINUTE_IN_MILLIS;
import static android.text.format.DateUtils.SECOND_IN_MILLIS;
import static se.jakob.knarkklocka.data.Alarm.STATE_ACTIVE;
import static se.jakob.knarkklocka.data.Alarm.STATE_DEAD;
import static se.jakob.knarkklocka.data.Alarm.STATE_SNOOZING;
import static se.jakob.knarkklocka.data.Alarm.STATE_WAITING;
import static se.jakob.knarkklocka.utils.TimerUtils.ACTION_STOP_ALARM;
import static se.jakob.knarkklocka.utils.TimerUtils.EXTRA_ALARM_ID;

public class AlarmActivity extends AppCompatActivity implements View.OnClickListener, View.OnLongClickListener {
    private static final String TAG = "AlarmActivity";   

    private Button mSnoozeButton;
    private Button mDismissButton;
    private TextView tv_alarm_text;

    private AlarmActivityViewModel alarmActivityViewModel;
 
    private Chronometer mAlarmChronometer;

    private boolean alarmIsActive = false;

    private AlarmManager.OnAlarmListener alarmCallback = new AlarmManager.OnAlarmListener() {
        @Override
        public void onAlarm() {
            if (BuildConfig.DEBUG) {
                Log.d(TAG, "Timeout reached. Snoozing...");
            }
            snooze();
        }
    };

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Log.d(TAG, "AlarmActivity started");
        setContentView(R.layout.activity_alarm);

        WakeLocker.acquire(this);

        /*Ensure that screen turns on*/
        final Window win = getWindow();
        if (Build.VERSION.SDK_INT >= 27) {
            setTurnScreenOn(true); //Replaces FLAG_TURN_SCREEN_ON
            setShowWhenLocked(true); //Replaces FLAG_SHOW_WHEN_LOCKED
            win.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON
                    | WindowManager.LayoutParams.FLAG_ALLOW_LOCK_WHILE_SCREEN_ON);
        } else {
            win.addFlags(WindowManager.LayoutParams.FLAG_TURN_SCREEN_ON);
            win.addFlags(WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED);
            win.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON
                    | WindowManager.LayoutParams.FLAG_ALLOW_LOCK_WHILE_SCREEN_ON);
        }

        /*Hide navigation bar*/
        hideNavigationBar();

        /* Close dialogs and window shade, so this is fully visible */
        sendBroadcast(new Intent(Intent.ACTION_CLOSE_SYSTEM_DIALOGS));

        mDismissButton = findViewById(R.id.dismiss);
        mSnoozeButton = findViewById(R.id.snooze);
        tv_alarm_text = findViewById(R.id.tv_alarm_text);

        mSnoozeButton.setOnClickListener(this);
        mDismissButton.setOnLongClickListener(this);

        mAlarmChronometer = findViewById(R.id.alarm_chronometer);

        /*Start timeout timer, so that alarm is not going off forever */
        startTimeoutClock();

        /*Get the ID of the Alarm that is going off*/
        Intent intent = getIntent();
        long id = intent.getLongExtra(EXTRA_ALARM_ID, -1);

        /*Setup ViewModel and Observer*/
        alarmActivityViewModel = ViewModelProviders.of(this).get(AlarmActivityViewModel.class);
        alarmActivityViewModel.setAlarm(id);
        alarmActivityViewModel.getAlarm().observe(this, new Observer<Alarm>() {
            @Override
            public void onChanged(@Nullable final Alarm alarm) {
                if (alarm != null) {
                    int state = alarm.getState();
                    switch (state) {
                        case STATE_ACTIVE:
                            startAlarm(alarm);
                            break;
                        case STATE_DEAD:
                            finish();
                            break;
                        case STATE_SNOOZING:
                            finish();
                            break;
                        case STATE_WAITING:
                            finish();
                            break;
                    }
                } else {
                    mAlarmChronometer.setVisibility(View.INVISIBLE);
                    tv_alarm_text.setVisibility(View.INVISIBLE);
                }
            }
        });

    }

    private void setupChronometer(Alarm alarm) {
        mAlarmChronometer.setVisibility(View.VISIBLE);
        tv_alarm_text.setVisibility(View.VISIBLE);
        Date endTime = alarm.getEndTime();
        long timeDelta = endTime.getTime() - System.currentTimeMillis();
        mAlarmChronometer.setBase(SystemClock.elapsedRealtime() + timeDelta);
        mAlarmChronometer.setOnChronometerTickListener(
                new Chronometer.OnChronometerTickListener() {
                    public void onChronometerTick(Chronometer chronometer) {
                        int onColor = ContextCompat.getColor(getApplication(), R.color.colorAccent);
                        int offColor = ContextCompat.getColor(getApplication(), android.R.color.darker_gray);
                        int currentColor = tv_alarm_text.getCurrentTextColor();
                        if (currentColor == onColor) {
                            tv_alarm_text.setTextColor(offColor);
                        } else {
                            tv_alarm_text.setTextColor(onColor);
                        }
                    }
                }
        );
        mAlarmChronometer.start();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (alarmIsActive) {
            snooze();
        }
    }

    private void snooze() {
        if (alarmIsActive) {
            TimerUtils.startSnoozeTimer(this, alarmActivityViewModel);
            stopAlarm();
        }
        finish();
    }

    private void dismiss() {
        if (alarmIsActive) {
            alarmActivityViewModel.kill();
            TimerUtils.startMainTimer(this, alarmActivityViewModel);
            stopAlarm();
        }
        finish();
    }

    private void hideNavigationBar() {
        getWindow().getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                | View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY);
    }

    @Override
    public void onClick(View view) {
        if (view == mSnoozeButton) {
            snooze();
        } else if (view == mDismissButton) {
            dismiss();
        }

    }

    private void startAlarm(Alarm alarm) {
        setupChronometer(alarm);
        alarmIsActive = true;
    }

    private void stopAlarm() {
        stopTimeoutClock();
        alarmIsActive = false;
        alarmActivityViewModel.getAlarm().removeObservers(this);
        Intent alarmIntent = new Intent(this, AlarmService.class);
        //alarmIntent.putExtra(EXTRA_ALARM_ID, id);
        alarmIntent.setAction(ACTION_STOP_ALARM);
        startService(alarmIntent);
    }

    private void startTimeoutClock() {
        AlarmManager alarmManager = getSystemService(AlarmManager.class);
        long timeout;
        if (BuildConfig.DEBUG) {
            timeout = 10 * SECOND_IN_MILLIS;
        } else {
            timeout = 2 * MINUTE_IN_MILLIS;
        }

        if (alarmManager != null) {
            alarmManager.setExact(RTC_WAKEUP, System.currentTimeMillis() + timeout, "tag", alarmCallback, null);
        }
    }

    private void stopTimeoutClock() {
        AlarmManager alarmManager = getSystemService(AlarmManager.class);
        alarmManager.cancel(alarmCallback);
    }

    @Override
    public boolean onLongClick(View v) {
        if (v == mDismissButton) {
            dismiss();
        }
        return true;
    }
}

File: app/src/main/java/se/jakob/knarkklocka/AlarmActivity.kt
package se.jakob.knarkklocka

import android.app.AlarmManager
import android.app.AlarmManager.RTC_WAKEUP
import android.app.KeyguardManager
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.Build
import android.os.Bundle
import android.os.IBinder
import android.os.SystemClock
import android.text.format.DateUtils.MINUTE_IN_MILLIS
import android.text.format.DateUtils.SECOND_IN_MILLIS
import android.util.Log
import android.view.View
import android.view.WindowManager.LayoutParams.*
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import kotlinx.android.synthetic.main.activity_alarm.*
import se.jakob.knarkklocka.data.Alarm
import se.jakob.knarkklocka.data.AlarmState
import se.jakob.knarkklocka.ui.ControllerFragment
import se.jakob.knarkklocka.utils.*
import se.jakob.knarkklocka.utils.TimerUtils.EXTRA_ALARM_ID
import se.jakob.knarkklocka.viewmodels.AlarmActivityViewModel


class AlarmActivity : AppCompatActivity(), ControllerFragment.OnControllerEventListener {

    private lateinit var viewModel: AlarmActivityViewModel

    private var alarmIsActive = false
    private var alarmIsHandled = false
    private var mServiceBound: Boolean = false

    // Animation
    private lateinit var animBlink : Animation

    private val timeoutLength = 5 * MINUTE_IN_MILLIS

    private val alarmCallback = AlarmManager.OnAlarmListener {
        if (BuildConfig.DEBUG) {
            Log.d(TAG, "Timeout reached. Snoozing...")
        }
        finish()
    }

    private val connection = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName?, service: IBinder?) {
            Log.d(TAG, "Finished binding to AlarmService")
        }

        override fun onServiceDisconnected(name: ComponentName?) {
            Log.d(TAG, "Disconnected from AlarmService")
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Log.d(TAG, "AlarmActivity started")

        bindAlarmService()

        WakeLocker.acquire(this)

        /*Ensure that screen turns on*/
        if (Build.VERSION.SDK_INT >= 27) {
            setShowWhenLocked(true)    //Replaces FLAG_SHOW_WHEN_LOCKED
            setTurnScreenOn(true)      //Replaces FLAG_TURN_SCREEN_ON
            val keyguardManager = (getSystemService(Context.KEYGUARD_SERVICE) as KeyguardManager)
            keyguardManager.requestDismissKeyguard(this, null)
            window.addFlags(
                    FLAG_KEEP_SCREEN_ON
                            or FLAG_ALLOW_LOCK_WHILE_SCREEN_ON)
        } else {
            window.addFlags(
                    FLAG_SHOW_WHEN_LOCKED
                            or FLAG_TURN_SCREEN_ON
                            or FLAG_KEEP_SCREEN_ON
                            or FLAG_ALLOW_LOCK_WHILE_SCREEN_ON
                            or FLAG_DISMISS_KEYGUARD)
        }

        /*Hide navigation and status bar*/
        hideUIElements()

        /*Get the ID of the Alarm that is going off*/
        val id = intent.getLongExtra(EXTRA_ALARM_ID, -1)

        /*Setup ViewModel and Observer*/
        val factory = InjectorUtils.provideAlarmActivityViewModelFactory(this, id)
        viewModel = ViewModelProviders.of(this, factory).get(AlarmActivityViewModel::class.java)

        setContentView(R.layout.activity_alarm)

        /* Close dialogs and window shade, so this is fully visible */
        sendBroadcast(Intent(Intent.ACTION_CLOSE_SYSTEM_DIALOGS))

        if (BuildConfig.DEBUG) {
            button_miss_alarm.visibility = View.VISIBLE
            button_miss_alarm.setOnClickListener {
                finish()
            }
        } else {
            button_miss_alarm.visibility = View.INVISIBLE
        }


        /*Start timeout timer, so that alarm is not going off forever */
        startTimeoutClock()



        viewModel.liveAlarm.observe(this, Observer {alarm : Alarm? ->
            if (alarm != null) {
                if (alarm.state == AlarmState.STATE_ACTIVE) {
                    startAlarm(alarm)
                } else {
                    finish()
                }
            } else {
                alarm_chronometer.visibility = View.INVISIBLE
                tv_alarm_text.visibility = View.INVISIBLE
            }
        })

        // load the animation
        animBlink = AnimationUtils.loadAnimation(this,
                R.anim.blink)

        // set animation listener
        //animBlink.setAnimationListener(this)

        tv_alarm_text.startAnimation(animBlink)

    }

    override fun onControllerEvent(v: View, event: String) {
        when (event) {
            ACTION_RESTART_ALARM -> {
                alarmIsHandled = true
                if (alarmIsActive) {
                    dismiss()
                } else {
                    finish()
                }
            }
            ACTION_SNOOZE_ALARM -> {
                alarmIsHandled = true
                if (alarmIsActive) {
                    snooze()
                } else {
                    finish()
                }
            }
            ACTION_SLEEP -> {
                alarmIsHandled = true
                if (alarmIsActive) {
                    sleep()
                } else {
                    finish()
                }
            }
        }
    }

    private fun setupChronometer(alarm: Alarm) {
        alarm_chronometer.visibility = View.VISIBLE
        val endTime = alarm.endTime
        val timeDelta = endTime.time - System.currentTimeMillis()
        alarm_chronometer.base = SystemClock.elapsedRealtime() + timeDelta
        alarm_chronometer.start()
    }

    override fun onDestroy() {
        super.onDestroy()
        if (alarmIsHandled) {
            AlarmBroadcasts.broadcastAlarmHandled(this)
            stopAlarm()
            WakeLocker.release()
        }
        unbindAlarmService()
    }


    private fun snooze() {
        viewModel.getCurrentAlarm()?.let { alarm ->
            TimerUtils.startSnoozeTimer(this, alarm)
        }
        finish()
    }

    private fun dismiss() {
        viewModel.kill() /* Kill old timer */
        TimerUtils.startMainTimer(this) /* Start a new timer */
        finish()
    }

    private fun sleep() {
        viewModel.getCurrentAlarm()?.let { alarm ->
            TimerUtils.cancelAlarm(this, alarm.id) /* Cancel alarm in AlarmManager */
        }
        viewModel.sleep() /* Update alarm in DB */
        finish()
    }

    private fun hideUIElements() {
        window.decorView.systemUiVisibility = (
                View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                        or View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                        or View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY
                        or View.SYSTEM_UI_FLAG_FULLSCREEN
                        or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN)
        supportActionBar?.hide()
    }

    override fun onBackPressed() {
        snooze()
    }

    private fun startAlarm(alarm: Alarm) {
        setupChronometer(alarm)
        alarmIsActive = true
    }

    private fun stopAlarm() {
        stopTimeoutClock()
        alarmIsActive = false
        viewModel.liveAlarm.removeObservers(this)
        AlarmBroadcasts.broadcastStopAlarm(this)
    }

    private fun startTimeoutClock() {
        val alarmManager = getSystemService(AlarmManager::class.java)
        val timeout = if (BuildConfig.DEBUG) {
            10 * SECOND_IN_MILLIS
        } else {
            timeoutLength
        }

        alarmManager.setExact(RTC_WAKEUP, System.currentTimeMillis() + timeout, "tag", alarmCallback, null)
    }

    private fun stopTimeoutClock() {
        val alarmManager = getSystemService(AlarmManager::class.java)
        alarmManager.cancel(alarmCallback)
    }

    /**
     * Bind AlarmService if not yet bound.
     */
    private fun bindAlarmService() {
        if (!mServiceBound) {
            val intent = Intent(this, AlarmService::class.java)
            bindService(intent, connection, Context.BIND_AUTO_CREATE)
            mServiceBound = true
        }
    }

    /**
     * Unbind AlarmService if bound.
     */
    private fun unbindAlarmService() {
        if (mServiceBound) {
            unbindService(connection)
            mServiceBound = false
        }
    }

    companion object {
        private const val TAG = "AlarmActivity"
    }
}
